### PR TITLE
DenseMatrix I/O moved to blocking buffer

### PR DESCRIPTION
**Pull request by jbloom22**
I also added skipBytes for use with RowMatrix.readBlockMatrix

@cseed I still need to experiment on compression w.r.t. banded LD project, which should inform the choice of BDM BufferSpec further, but using the general infrastructure is clearly better than the one-off BDM buffers I'd written before. Is it reasonable to merge this to master (post review)?

### CODE REVIEW

File: src/main/scala/is/hail/utils/richUtils/RichDenseMatrixDouble.scala [hunk 1]
```diff
@@ -1,62 +1,27 @@
...
-      nLeft -= nCopy
-    }
-  }
-
-  // assumes zero offset and minimal majorStride 
-  def read(dis: DataInputStream): DenseMatrix[Double] = {
-    val rows = dis.readInt()
-    val cols = dis.readInt()
-    val isTranspose = dis.readBoolean()
+  val bufferSpec = CodecSpec.blockSpecs(1)
```
Comment (cseed) [inline src/main/scala/is/hail/utils/richUtils/RichDenseMatrixDouble.scala]: Don't use `blockSpecs`, those are really just for testing (but used for python so I couldn't put them in the tests area).

The buffer spec needs to be written to the metadata and that's the spec you need to use on reading.  That lets you change the buffer spec in a forward compatible way (e.g. add a double dictionary encoding buffer as Richard suggested).  See e.g. Table.read/write for examples.

Comment (cseed) [reply src/main/scala/is/hail/utils/richUtils/RichDenseMatrixDouble.scala]: Also, slightly strange place to put this.  It should originate in the write function and be passed around from there.

**Comment by jbloom22** [comment]
You can push back on the extra flush to align the block row with the buffer. It shouldn't matter too much either way, but to the extent there is redundancy per row (as there will be in UKBB LD case) I figure alignment is a net positive.

File: src/main/scala/is/hail/linalg/BlockMatrix.scala [hunk 2]
```diff
@@ -1149,20 +1150,22 @@ class WriteBlocksRDD(path: String,
...
+          outPerBlockCol(blockCol).writeDoubles(data, 0, n)          
           blockCol += 1
         }
         i += 1
       }
     }
 
-    dosPerBlockCol.foreach(_.close())
+    outPerBlockCol.foreach { out =>
+      out.flush()
```
Comment (cseed) [inline src/main/scala/is/hail/linalg/BlockMatrix.scala]: Not required before close.

File: src/main/scala/is/hail/linalg/BlockMatrix.scala [hunk 3]
```diff
@@ -1100,22 +1098,23 @@ class WriteBlocksRDD(path: String,
...
-      dos.writeBoolean(true) // transposed, stored row major
-
-      dos
+      val os = sHadoopBc.value.value.unsafeWriter(filename)
+      val out = BlockMatrix.bufferSpec.buildOutputBuffer(os)
+      
+      out.writeInt(nRowsInBlock)
+      out.writeInt(nColsInBlock)
+      out.writeBoolean(true) // transposed, stored row major
+      out.flush() // align block row with buffer
```
Comment (cseed) [inline src/main/scala/is/hail/linalg/BlockMatrix.scala]: Yeah, I'm against this.  I don't think it buys you anything but makes the code more brittle.

File: src/main/scala/is/hail/utils/richUtils/RichDenseMatrixDouble.scala [hunk 4]
```diff
@@ -135,19 +98,22 @@ class RichDenseMatrixDouble(val m: DenseMatrix[Double]) extends AnyVal {
...
-    dos.writeInt(m.rows)
-    dos.writeInt(m.cols)
-    dos.writeBoolean(isTranspose)
+    val out = bufferSpec.buildOutputBuffer(os)
     
-    RichDenseMatrixDouble.writeDoubles(dos, data, data.length)
+    out.writeInt(m.rows)
+    out.writeInt(m.cols)
+    out.writeBoolean(isTranspose)
+    out.flush() // align block row with buffer
```
Comment (cseed) [inline src/main/scala/is/hail/utils/richUtils/RichDenseMatrixDouble.scala]: Same.

**Comment by cseed** [comment]
Any size data yet?

**Pull request merged**